### DISCUSSION TITLE
Adding heat source and temperature dependent material properties with navier stokes

### DISCUSSION THREAD

**Discussion by Jake-HW-Miles** [idaholab/moose#27397] [Q&A Modules: Navier-Stokes] [answered]
Hi all, 

I am trying to simulate fluid flow, based on the thesis attached. So far, I have managed to successfully get it working, however the next step is to add a volumetric heat source function which varies with time. The function is currently defined as a ParsedFunction.

I am using INSFV implementation and can provide more detail if needed. I was previously using the HeatSource Kernel when not using Navier Stokes, so want to know if there is an equivalent. 

I tried using a portion of the NavierStokesFV module, however no luck...

My attempt:
```
[Modules]
    [NavierStokesFV]
        pin_pressure = true
        pinned_pressure_type = average-uo
        external_heat_source = volumetric_decay_func
        block = fuel_salt
        density = ${rho}
        dynamic_viscosity = ${mu}
       []
[]

```
If this is a trivial solve, the second part of my question pertains to using functions to alter material properties, such as density, with temperature which seemingly requires FunctorMaterial properties.

Happy to expand on this! Thank you in advance!

Thesis: Verification of the fluid dynamics modules of the Multiphysics simulation framework MOOSE
Link: https://www.diva-portal.org/smash/get/diva2:1652721/FULLTEXT01.pdf

**Comment by GiudGiud** [comment, marked as answer]
Hi Jake

A heat source should be easy to add.
If you have a working Modules/NavierStokes action syntax, adding the external_heat_source is the correct thing to do.
One problem may be if you are trying to block restrict the definition of the source? (not apply heat over the whole domain).
In that case, you can still use a `FVBodyForce` FVKernel on the side to define this.

For material properties, you will want to use a GeneralFunctorFluidProps material. 
Many of the fluid properties will work without extra work in NavierStokes. 
https://mooseframework.inl.gov/syntax/FluidProperties/index.html

Do you have a working input without a heat source right now?

Guillaume

**Reply by Jake-HW-Miles** [reply]
Hi Guillaume,

Thank you for the reply, I wanted to give these things a go before getting back to you. I have (partially) succeeded in implementing these suggestions.

FYI I am not using the Navier Stokes module at this stage, and am instead using the respective FVKernels.

- `FVBodyForce` was exactly what was needed for the implementation of a heat source in a specified block. 
- `GeneralFunctorFluidProps` was also what was required, where my fluid properties are now defined using `TemperaturePressureFunctionFluidProperties`.

My only follow up is due to confusion regarding the use of the functor properties given by `GeneralFunctorFluidProps`:

Currently this is my `GeneralFunctorFluidProps` block:
```
[fuel_salt]
    type = GeneralFunctorFluidProps
    block = fuel_salt
    fp = salt_fluid_properties

    pressure = pressure
    T_fluid = T
    characteristic_length = 0.2
    speed = velocity_xy

    force_define_density = true
[]
```

Which I believe is correct, as it has no issues being read in all my FVKernels, except one...
It does not work for `FVHeatConductionTimeDerivative`, where density_name is requested instead of rho, like the other FVKernels I am using for NS. I have tried simply defining an `ADParsedMaterial` which defines a new density, which should vary the same. However my follow up question is whether I can do this in a different way?

Hope that makes sense! Thank you for this advice!

Jake

**Reply by GiudGiud** [reply]
Hello

For `FVHeatConductionTimeDerivative` you would be able to set: `density_name = rho`
except not, because that kernel uses regular material properties instead of functor material properties.

Why do you want to use that kernel? The (P)INSFVEnergyTimeDerivative are meant to be used with NSFV, for both fluid and solid time derivatives.
The reason for the distinction is actually the use of a different material property system

Another way (same as the ADParsedMaterial) is to use this converter between systems
https://mooseframework.inl.gov/source/materials/MaterialFunctorConverter.html
that's if you really need to use that kernel

Guillaume

**Reply by Jake-HW-Miles** [reply]
Hi!
The use of `FVHeatConductionTimeDerivative` is my mistake, I thought I had seen it in a INSFV example... Upon checking the example* , it seems not! I cannot see why I would need it at this stage.

Thank you for providing a clear and concise response. I will mark your first response as the answer. 

Best regards,
Jake

* modules/navier_stokes/test/tests/postprocessors/rayleigh/natural_convection.i
